File: python/problem_13.py
def longest_substring_k_distinct(s: str, k: int) -> int:
    # Uses a sliding window with a character-frequency dict.
    # Expand right; shrink left whenever distinct chars exceed k.
    # This runs in O(n) time and O(k) space.
    freq: dict[str, int] = {}
    left = best = 0

    for right, ch in enumerate(s):
        freq[ch] = freq.get(ch, 0) + 1  # expand window

        # shrink from left until at most k distinct chars remain
        while len(freq) > k:
            lch = s[left]
            freq[lch] -= 1
            if freq[lch] == 0:
                del freq[lch]
            left += 1

        best = max(best, right - left + 1)

    return best


def main():
    print("Problem 13:")
    assert longest_substring_k_distinct("abcba", 2) == 3, "assertion failed"
    assert longest_substring_k_distinct("aa", 1)    == 2, "assertion failed"
    assert longest_substring_k_distinct("abc", 3)   == 3, "assertion failed"
    assert longest_substring_k_distinct("", 2)      == 0, "assertion failed"
    print("All Tests Passed")

File: c/problem_13.c
#include "problem_13.h"
#include <stdio.h>
#include <string.h>
#include <assert.h>

/* longestKDistinct: sliding window with frequency array. O(n) time. */
int longestKDistinct(const char *s, int k) {
    int freq[256] = {0};
    int distinct = 0, left = 0, best = 0;
    int n = (int)strlen(s);

    for (int right = 0; right < n; right++) {
        unsigned char ch = (unsigned char)s[right];
        if (freq[ch] == 0) distinct++;
        freq[ch]++;
        while (distinct > k) {
            unsigned char lch = (unsigned char)s[left++];
            if (--freq[lch] == 0) distinct--;
        }
        if (right - left + 1 > best) best = right - left + 1;
    }
    return best;
}

void problem13(void) {
    printf("Problem 13:\n");
    assert(longestKDistinct("abcba", 2) == 3);
    assert(longestKDistinct("aa",    1) == 2);
    assert(longestKDistinct("abc",   3) == 3);
    assert(longestKDistinct("",      2) == 0);
    printf("All Tests Passed\n");
}
